File: Segundo periodo/AED-2/TP01-parte1/PalindromoRecursivo.java
/*
 * Autor: Luiz Junio <luisjuniorbr@gmail.com>
 * Criado em: 23/08/2018
 * Enunciado:
 * Palindromo - Crie um metodo *recursivo que recebe uma string como parametro e retorna true
 * se essa e um palindromo. Na saida padrao, para cada linha de entrada, escreva uma linha de
 * saida com SIM / NAO indicando se a linha e um palindromo.
 */

public class PalindromoRecursivo
{//Inicio classe PalindromoRecursivo
    public static void main(String[] args)
    {//Inicio main
        String[] palavra = new String[1000];
        int numEntrada = 0;

        do{//Leitura da entrada padrao  
            palavra[numEntrada] = MyIO.readLine();
        }while (!ehIgual(palavra[numEntrada++],"FIM"));

        numEntrada--;   //Desconsiderar ultima linha contendo a palavra FIM

        //Impressao da saida
        for(int i = 0; i < numEntrada; i++)
        {//Inicio for
            if(ehPalindromo(palavra[i])){
                MyIO.println("SIM");
            }
            else MyIO.println("NAO");
        }//Fim for    

    }//Fim main

    public static boolean ehPalindromo(String palavra)
    {//Inicio ehPalindromo
        return ehPalindromo(palavra, 0);
    }//Fim ehPalindromo

    private static boolean ehPalindromo(String palavra, int i)
    {//Inicio ehPalindromo
        boolean palindromo = true;
        int ultimo = palavra.length() - 1;
        if(i < palavra.length() / 2 ){
            if(palavra.charAt(i) == palavra.charAt(ultimo - i))
                palindromo = true && ehPalindromo(palavra, i + 1);
            else
                palindromo = false;
        }
        return palindromo;
    }//Fim ehPalindromo

    public static boolean ehIgual(String p1, String p2)
    {//Inicio ehIgual
        boolean resp = true;
        if(p1.length() != p2.length())
            resp = false;
        else
            resp = ehIgual(p1,p2,0);
        return resp;
    }//Fim ehIgual

    private static boolean ehIgual(String p1, String p2, int i)
    {//Inicio ehIgual
     /***************************************************************
      * Nome do método: ehIgual
      * Data da elaboração: 23/08/2018
      * Data da última alteração: 23/08/2018
      * Autor: Luiz Junio <luisjuniorbr@gmail.com>
      * Contexto de ação: Verifica se uma String é igual a outra.
      * Valor gerado: verdadeiro se forem iguais, falso caso contrário.
      ****************************************************************
      */ 
        boolean igual = true;
        if(i < p1.length()){
            if(p1.charAt(i) == p2.charAt(i))
                igual = true && ehIgual(p1,p2,i + 1);
            else
                igual = false;
        }
        return igual;
    }//Fim ehIgual

}//Fim classe PalindromoRecursivo

File: Segundo periodo/AED-2/TP01-parte1/PalindromoRecursivo.c
/*
 *  Autor: Luiz Junio <luisjuniorbr@gmail.com>
 *  Enunciado:
 *   Palindromo em C - Na linguagem C, crie um metodo *Recursivo que recebe uma string como
 *   parametro e retorna true se essa e um palindromo. Na saida padrao, para cada linha de entrada,
 *   escreva uma linha de saida com SIM / NAO indicando se a linha e um palindromo.
 *   Observacao: Uma linha de entrada pode ter caracteres nao letras.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int ehPalindromo(char palavra[], int i)
{//Inicio ehPalindromo
    /***************************************************************
    * Nome do metodo: ehPalindromo
    * Data da elaboração: 14/08/2018
    * Data da ultima alteração: 14/08/2018
    * Autor: Luiz Junio <luisjuniorbr@gmail.com>
    * Contexto de ação: Verificar se uma String pode ser lida da esquerda
    * para a direita, e vice-versa.
    * Valor gerado: 1 se for um paindromo, 0 caso contrário.
    ****************************************************************
    */
    int palindromo = 1;//TRUE
    int ultimo = strlen(palavra) - 1;//Posicao da ultima letra

    if(i < strlen(palavra) / 2)
    {//Inicio if
        if(palavra[i] == palavra[ultimo - i]){
            palindromo = ehPalindromo(palavra, i + 1);
        }
        else{
            palindromo = 0;//FALSE
        }
    }//Fim if
    return palindromo;
}//Fim ehPalindromo

int main()
{//Inicio main
    char palavra[1000][1000];//Iniciando um Array de "Strings"
    int numEntrada = 0;
    do{
        //Esse scanf irá ler a string, incluindo espaços.
        //O espaço antes de % serve para limpar o buffer do teclado 
        scanf(" %[^\n]s", palavra[numEntrada]);
    }while(strcmp(palavra[numEntrada++], "FIM") != 0);//Repete enquanto não for igual a FIM

    numEntrada--; //Desconsiderar ultima linha contendo a palavra FIM

    for(int i = 0; i < numEntrada; i++)
    {//Inicio for
        if(ehPalindromo(palavra[i], 0)){
            printf("SIM\n");
        }
        else{
            printf("NAO\n");
        }
    }//Fim for
    return 0;//Indica que o programa terminou com sucesso
}//Fim main 

